metricbeat/mb/mb.go

```diff
@@ -61,9 +63,17 @@ const (
 
 // Module is the common interface for all Module implementations.
 type Module interface {
-	Name() string                      // Name returns the name of the Module.
-	Config() ModuleConfig              // Config returns the ModuleConfig used to create the Module.
-	UnpackConfig(to interface{}) error // UnpackConfig unpacks the raw module config to the given object.
+	// Name returns the name of the Module.
+	Name() string
+
+	// Config returns the ModuleConfig used to create the Module.
+	Config() ModuleConfig
+
+	// UnpackConfig unpacks the raw module config to the given object.
+	UnpackConfig(to interface{}) error
+
+	// ReConfigure reconfigures the module with the new configuration object.
+	ReConfigure(config *common.Config, register *Register) error
```

Comment:
Sure. I'd like to learn — mostly for my own curiousity — why this difference is significant. The only place I could find in the Metricbeat codebase where `mb.Module` is implemented but `mb.BaseModule` is not embedded was here: https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/metricbeat/mb/module/wrapper.go#L53.

Reply:
Changed in fc45f1d55.

Reply:
> Sure. I'd like to learn — mostly for my own curiousity — why this difference is significant. The only place I could find in the Metricbeat codebase where mb.Module is implemented but mb.BaseModule is not embedded was here: 

Well, this is not so significant, but there are two reasons why I would prefer to avoid adding a method to this interface:
* `mb.Module` is something like a "read-only" interface, that can be used to access to information about the module, but not modifying it, so `mb.Module` can be used as a way to control visibility. When a method receives a `mb.Module` it knows that it is only to get information of the module. When a method receives a `*mb.BaseModule` it has access to all the base module, this can be intended for methods like the module builders, as the module is still being built there, it can make sense to allow modifications of the module.
* `mb.Module` is a public interface, adding methods to it can break existing code (as happened with the test [here](https://github.com/elastic/beats/pull/17609/files/d1214273d5b267fb7bfba554d8707274100e156e#r406092724)). Though I don't think many community beats are implementing their own mb.Modules... so not so important.